docs/blueprints/referral-manager-v1.md

```diff
@@ -0,0 +1,236 @@
+# Referral Manager V1: Cross-Satellite Referral Link
+
+## Abstract
+
+This document describes how to handle user referrals across satellites.
+
+## Background
+
+Previous referral program only works per satellite. This means a user on us-east-1 satellite can only refer people to join us-east-1.
+
+We want referrals to be valid on all Tardigrade satellites. This will allow the user to select the satellite that is best for them.
+
+### Non-Goals
+A registration page where a user can select their preferred Satellite during registration.
+
+## Design
+
+### Definitions
+
+Referral Manager
+: A stand-alone process running outside of satellites. It should only talk to tardigrade branded satellites.
+
+Referral Manager CLI
+: A CLI that will allow operators to invoke referral link generation.
+
+Referral Link
+: A one-time-use URL that contains a unique invitation token.
+
+Invitation Token
+: A random number generated by Referral Manager for each user in the referral program.
+
+Eligible Users
+: Users who have 0 unredeemed referral link.
+
+Owner's Satellite
+: The satellite that an owner of a particular token belongs to.
+
+Redeemed Satellite
+: The satellite where the new user redeeming a token has registered on.
+
+### How it will work
+
+_User Interface_
+
+1. Users will have a referral tab on the UI. On click, it will trigger satellite to send a request to retrieve referral links from Referral Manager:
+    - Referral Manager will query unredeemed tokens in `tokens` table for this user ID and satellite.
+        - if there are unredeemed tokens, Referral Manager respond satellite's request with the tokens.
+        - if no unredeemed token is found for this user ID and such user ID doesn't exist in `users` table, Referral Manager will add an entry for this user ID in `users` table.
+    - Satellite receives the response from Referral Manager.
+        - if the response playload contains tokens, satellite will display them in the UI.
+        - if its an empty response payload, satellite will display a message `No available referral link. Try again later.`
+
+
+_Referral Manager CLI_
+
+1. `referral manager start` initiates the invitation token generation process. This process begins by checking whether there are existing `eligibleUsers` in memory in Referral Manager.
```

Comment:
we should just use the database. no sense in having a list in memory and in the database. just query the database for eligible users directly and combine these eligibleUsers and users query steps

Reply:
specifically, the cli should just find eligible users and create them tokens and store them back in the database. if that happens in a transaction it's okay to run multiple copies of the CLI together

Reply:
In addition, having things in memory would horizontal scaling more difficult.

Reply:
How would the CLI user know how many tokens they want to create per user if they don't know how many total `eligibleUsers` there are? Or do we just want them to input the total number of tokens they want to create? If that's the case, do we care to make the selected `eligibleUsers` to be evenly distributed across all satellites?

Reply:
Fixed